sdk/src/trace/span.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "src/trace/tracer.h"
+
+#include <mutex>
+
+namespace opentelemetry
+{
+namespace sdk
+{
+namespace trace
+{
+namespace trace_api = opentelemetry::trace;
+
+class Span final : public trace_api::Span
+{
+public:
+  explicit Span(std::shared_ptr<Tracer> &&tracer,
+                nostd::string_view name,
+                const trace_api::StartSpanOptions &options) noexcept;
+
+  ~Span() override;
+
+  void AddEvent(nostd::string_view name) noexcept override;
+
+  void AddEvent(nostd::string_view name, core::SystemTimestamp timestamp) noexcept override;
+  void AddEvent(nostd::string_view name, core::SteadyTimestamp timestamp) noexcept override;
+
+  void SetStatus(trace_api::CanonicalCode code, nostd::string_view description) noexcept override;
+
+  void UpdateName(nostd::string_view name) noexcept override;
+
+  void End() noexcept override;
+
+  bool IsRecording() const noexcept override;
+
+  trace_api::Tracer &tracer() const noexcept override { return *tracer_; }
+
+private:
+  std::shared_ptr<Tracer> tracer_;
+  std::mutex mutex_;
+  std::unique_ptr<Recordable> recordable_;
```

Comment:
`GUARDED_BY(mutex_)`?

Reply:
I'd need to define that macro somewhere, right?

Reply:
Yeah. I'd be happy to do this in a followup PR. In general, do you like `-Wthread-safety`?

https://github.com/abseil/abseil-cpp/blob/2f76a9bf50046e396138cc8eeb3cdc17b7a5ac24/absl/base/internal/thread_annotations.h#L64

Reply:
That looks good. But we should be careful so that they don't clash with a project that might already have defined those macros.

Maybe only define then within the src directory? or check if they're already defined?

Reply:
Yeah specific to the SDK and wrap it in `#ifndef`

Happy to do this in a followup PR.

I have no idea why `thread_annotations.h` doesn't come with Clang [last time I looked].